lib/iris/fileformats/netcdf/loader.py

```diff
@@ -0,0 +1,596 @@
+    #  into standard behaviour.
+    # Deferred import to avoid circular imports.
+    from iris.experimental.ugrid.cf import CFUGridReader
+    from iris.experimental.ugrid.load import (
+        PARSE_UGRID_ON_LOAD,
+        _build_mesh_coords,
+        _meshes_from_cf,
+    )
+    from iris.io import run_callback
+
+    # Create a low-level data-var filter from the original load constraints, if they are suitable.
+    var_callback = _translate_constraints_to_var_callback(constraints)
+
+    # Create an actions engine.
+    engine = _actions_engine()
+
+    if isinstance(filenames, str):
+        filenames = [filenames]
+
+    for filename in filenames:
+        # Ingest the netCDF file.
+        meshes = {}
+        if PARSE_UGRID_ON_LOAD:
+            cf = CFUGridReader(filename)
+            meshes = _meshes_from_cf(cf)
+        else:
+            cf = iris.fileformats.cf.CFReader(filename)
+
+        # Process each CF data variable.
+        data_variables = list(cf.cf_group.data_variables.values()) + list(
+            cf.cf_group.promoted.values()
+        )
+        for cf_var in data_variables:
+            if var_callback and not var_callback(cf_var):
+                # Deliver only selected results.
+                continue
+
+            # cf_var-specific mesh handling, if a mesh is present.
+            # Build the mesh_coords *before* loading the cube - avoids
+            # mesh-related attributes being picked up by
+            # _add_unused_attributes().
+            mesh_name = None
+            mesh = None
+            mesh_coords, mesh_dim = [], None
+            if PARSE_UGRID_ON_LOAD:
+                mesh_name = getattr(cf_var, "mesh", None)
+            if mesh_name is not None:
+                try:
+                    mesh = meshes[mesh_name]
+                except KeyError:
+                    message = (
+                        f"File does not contain mesh: '{mesh_name}' - "
+                        f"referenced by variable: '{cf_var.cf_name}' ."
+                    )
+                    logger.debug(message)
+            if mesh is not None:
+                mesh_coords, mesh_dim = _build_mesh_coords(mesh, cf_var)
+
+            cube = _load_cube(engine, cf, cf_var, filename)
+
+            # Attach the mesh (if present) to the cube.
+            for mesh_coord in mesh_coords:
+                cube.add_aux_coord(mesh_coord, mesh_dim)
+
+            # Process any associated formula terms and attach
+            # the corresponding AuxCoordFactory.
+            try:
+                _load_aux_factory(engine, cube)
+            except ValueError as e:
+                warnings.warn("{}".format(e))
+
+            # Perform any user registered callback function.
+            cube = run_callback(callback, cube, cf_var, filename)
+
+            # Callback mechanism may return None, which must not be yielded
+            if cube is None:
+                continue
+
+            yield cube
```

Comment:
?? this code has not changed from [the original](https://github.com/SciTools/iris/blob/56a97d66b8404f96ef61fe2d52433eb58373f338/lib/iris/fileformats/netcdf.py#L992-L996)

The "yield" just means that the routine is a generator, which provides an "iterator over cubes" rather than returning an "iterable of cubes"   -- i.e. it only calculates + delivers one at a time.
See : https://docs.python.org/3/glossary.html#term-generator
Generators are superior to functions returning an iterable because you don't need to calculate all the answers in advance, which saves time+memory.  You can even produce an infinite sequence, and have the caller choose when to stop.

Reply:
Some confusion on my part. Happy that this is all fine.